File: android/app/src/test/kotlin/run/nuri/getagrip/MockProgressorClientTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.advanceTimeBy
import kotlinx.coroutines.test.runCurrent
import kotlinx.coroutines.test.runTest
import run.nuri.getagrip.ble.MockProgressorClient
import run.nuri.getagrip.ble.StreamStartCause
import run.nuri.getagrip.engine.ForceSample
import run.nuri.getagrip.engine.ProgressorCommand
import run.nuri.getagrip.engine.ProgressorEvent
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

@OptIn(ExperimentalCoroutinesApi::class)
class MockProgressorClientTests {
    @Test fun repeatedTareZeroesTheCurrentLoadWithoutAccumulatingOffsets() = runTest {
        val client = MockProgressorClient(backgroundScope, clock = FakeClock())
        val samples = mutableListOf<ForceSample>()
        client.onEvent = { if (it is ProgressorEvent.Sample) samples += it.sample }
        client.connect()
        advanceTimeBy(450)
        runCurrent()
        assertTrue(client.state.isConnected)
        client.startStreaming(StreamStartCause.initial)
        advanceTimeBy(100)
        runCurrent()
        assertTrue(samples.last().kg > 1)

        val next = samples.size
        client.send(ProgressorCommand.tare)
        client.send(ProgressorCommand.tare)
        client.startStreaming(StreamStartCause.tareRecovery)
        advanceTimeBy(100)
        runCurrent()

        assertEquals(0.0, samples[next].kg, 1e-9,
            "Taring twice against one raw load must still mean zero")
        assertEquals(0u, samples[next].deviceMicros, "A real re-kick may reset the device epoch")
        samples.forEachIndexed { index, sample -> assertEquals(index % 8 == 0, sample.isBatchStart) }
        client.disconnect()
    }

    @Test fun secondSessionStartsUnloadedAfterEndingOnTheLoadedProfile() = runTest {
        val client = MockProgressorClient(backgroundScope, clock = FakeClock())
        val samples = mutableListOf<ForceSample>()
        client.onEvent = { if (it is ProgressorEvent.Sample) samples += it.sample }
        client.connect()
        advanceTimeBy(450)
        runCurrent()
        client.send(ProgressorCommand.tare)
        client.startStreaming(StreamStartCause.initial)
        advanceTimeBy(100)
        runCurrent()
        assertTrue(samples.last().kg > 1)
        client.send(ProgressorCommand.stopWeightMeasurement)

        val next = samples.size
        client.send(ProgressorCommand.tare)
        client.startStreaming(StreamStartCause.initial)
        runCurrent()

        assertEquals(0.0, samples[next].kg, 1e-9)
        assertEquals(0u, samples[next].deviceMicros)
        assertTrue(samples.drop(next).all { it.kg >= 0 },
            "The old synthetic load must not become a negative offset")
        client.disconnect()
    }
}

File: Tests/MockProgressorClientTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

@MainActor
final class MockProgressorClientTests: XCTestCase {
    private func waitUntil(_ condition: @MainActor () -> Bool) async throws {
        let deadline = ContinuousClock.now + .seconds(3)
        while !condition(), ContinuousClock.now < deadline {
            try await Task.sleep(for: .milliseconds(10))
        }
        XCTAssertTrue(condition(), "Mock did not deliver the expected connection or samples")
    }

    func testRepeatedTareZeroesTheCurrentLoadWithoutAccumulatingOffsets() async throws {
        let client = MockProgressorClient()
        var samples: [ForceSample] = []
        client.onEvent = { if case .sample(let sample) = $0 { samples.append(sample) } }
        client.connect()
        defer { client.disconnect() }
        try await waitUntil { client.state.isConnected }
        client.startStreaming(cause: .initial)
        try await waitUntil { samples.count >= 16 }
        XCTAssertGreaterThan(samples.last!.kg, 1)

        let next = samples.count
        client.send(.tare)
        client.send(.tare)
        client.startStreaming(cause: .tareRecovery)
        try await waitUntil { samples.count > next }

        XCTAssertEqual(samples[next].kg, 0, accuracy: 1e-9,
                       "Taring twice against one raw load must still mean zero")
        XCTAssertEqual(samples[next].deviceMicros, 0, "A real stream re-kick may reset the device epoch")
        for (index, sample) in samples.enumerated() { XCTAssertEqual(sample.isBatchStart, index % 8 == 0) }
    }

    func testSecondSessionStartsUnloadedAfterEndingOnTheLoadedProfile() async throws {
        let client = MockProgressorClient()
        var samples: [ForceSample] = []
        client.onEvent = { if case .sample(let sample) = $0 { samples.append(sample) } }
        client.connect()
        defer { client.disconnect() }
        try await waitUntil { client.state.isConnected }
        client.send(.tare)
        client.startStreaming(cause: .initial)
        try await waitUntil { samples.count >= 16 }
        XCTAssertGreaterThan(samples.last!.kg, 1)
        client.send(.stopWeightMeasurement)

        let next = samples.count
        // RunnerSession tares before starting each new session, without disconnecting.
        client.send(.tare)
        client.startStreaming(cause: .initial)
        try await waitUntil { samples.count > next }

        XCTAssertEqual(samples[next].kg, 0, accuracy: 1e-9)
        XCTAssertEqual(samples[next].deviceMicros, 0)
        XCTAssertTrue(samples[next...].allSatisfy { $0.kg >= 0 },
                      "The old synthetic load must not become a negative offset")
    }
}
